internal/cluster/store/store.go

```diff
@@ -0,0 +1,855 @@
+	stats.Add(snapshotCreateDuration, 0)
+	stats.Add(numSnapshotPersists, 0)
+	stats.Add(numSnapshotPersistsFailed, 0)
+	stats.Add(snapshotPersistDuration, 0)
+	stats.Add(leaderChangesObserved, 0)
+	stats.Add(leaderChangesDropped, 0)
+	stats.Add(failedHeartbeatObserved, 0)
+	stats.Add(nodesReapedOK, 0)
+	stats.Add(nodesReapedFailed, 0)
+}
+
+type Store struct {
+	config Config
+
+	open *atomic.Bool
+
+	notifyMu        sync.Mutex
+	bootstrapExpect int
+	bootstrapped    bool
+	notifyingNodes  map[string]raft.Server
+
+	mu     sync.Mutex
+	boltDB *raftboltdb.BoltStore
+
+	layer  *tcp.Layer
+	raftTn *raft.NetworkTransport
+
+	raftID    string // Node ID.
+	raftDir   string
+	raft      *raft.Raft // The consensus mechanism
+	boltStore *raftboltdb.BoltStore
+	logger    hclog.Logger
+
+	// Raft changes observer
+	observer      *raft.Observer
+	observerChan  chan raft.Observation
+	observerClose chan struct{}
+	observerDone  chan struct{}
+
+	state ClusterState
+}
+
+type Config struct {
+	NodeId              string
+	RetainSnapshotCount int
+	RaftTimeout         time.Duration
+	RaftDir             string
+
+	// Node-reaping configuration
+	ReapTimeout         time.Duration
+	ReapReadOnlyTimeout time.Duration
+
+	// Time after which the node becomes marked as shut down when it stops sending heartbeats
+	// must be lower than reap timeout if set
+	NodeHearbeatShutdownTimeout time.Duration
+}
+
+// DefaultConfig provides default store configuration based on cluster configuration.
+func DefaultConfig(c config.Cluster) Config {
+	conf := Config{
+		RetainSnapshotCount:         2,
+		RaftDir:                     c.RaftDir,
+		ReapTimeout:                 0,
+		RaftTimeout:                 5 * time.Second,
+		ReapReadOnlyTimeout:         0,
+		NodeId:                      c.NodeId,
+		NodeHearbeatShutdownTimeout: 2 * time.Second,
+	}
+	if c.RaftDir == "" {
+		conf.RaftDir = "zenbpm_raft"
+	}
+
+	if c.NodeId == "" {
+		conf.NodeId = random.String()
+	}
+	return conf
+}
+
+// New returns a new Store.
+// The store is in closed state and needs to be opened by calling Open before usage.
+func New(layer *tcp.Layer, c Config) *Store {
+	s := &Store{
+		config:  c,
+		mu:      sync.Mutex{},
+		boltDB:  &raftboltdb.BoltStore{},
+		raft:    &raft.Raft{},
+		logger:  hclog.Default().Named("zenbpm-store"),
+		open:    &atomic.Bool{},
+		raftID:  c.NodeId,
+		layer:   layer,
+		raftDir: c.RaftDir,
+		state: ClusterState{
+			Partitions: map[uint32]Partition{},
+			Nodes:      map[string]Node{},
+		},
+	}
+
+	ResetStats()
+	return s
+}
+
+// Open opens the store and configures underlying raft communication and storage
+func (s *Store) Open() (retErr error) {
+	defer func() {
+		if retErr == nil {
+			s.open.Store(true)
+		}
+	}()
+
+	if s.open.Load() {
+		return ErrAlreadyOpen
+	}
+	s.logger.Info(fmt.Sprintf("opening store with node ID %s, listening on %s", s.raftID, s.layer.Addr().String()))
+
+	// Setup Raft configuration.
+	config := raft.DefaultConfig()
```

Comment:
variable config collides with imported package name
I recommend renaming the variable e.g. `cfg`